File: packages/contracts-python/anvilkit_contracts/validator.py
"""Pinned Python PLAN-0003 M4 validation adapter."""

from __future__ import annotations

import hashlib
import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator, FormatChecker
from referencing import Registry, Resource

MAX_SAFE_INTEGER = 9_007_199_254_740_991


@dataclass(frozen=True, order=True)
class Finding:
    code: str
    instance_path: str
    schema_path: str


class AdmissionError(ValueError):
    pass


def _pairs(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise AdmissionError("duplicate decoded key")
        result[key] = value
    return result


def _integer(wire: str) -> int:
    value = int(wire)
    if abs(value) > MAX_SAFE_INTEGER:
        raise AdmissionError("unsafe integer")
    if wire.startswith("-") and value == 0:
        raise AdmissionError("negative zero")
    return value


def _number(wire: str) -> float:
    value = float(wire)
    if not math.isfinite(value) or (value == 0 and wire.startswith("-")):
        raise AdmissionError("number range or negative zero")
    return value


def _limits(value: Any, depth: int = 1) -> int:
    if depth > 64:
        raise AdmissionError("depth limit")
    if isinstance(value, str):
        if any(0xD800 <= ord(character) <= 0xDFFF for character in value):
            raise AdmissionError("invalid Unicode scalar")
        if len(value.encode("utf-8")) > 1_048_576:
            raise AdmissionError("string byte limit")
        return 1
    if isinstance(value, list):
        if len(value) > 100_000:
            raise AdmissionError("item limit")
        return 1 + sum(_limits(item, depth + 1) for item in value)
    if isinstance(value, dict):
        if len(value) > 100_000:
            raise AdmissionError("item limit")
        return 1 + sum(_limits(key, depth + 1) + _limits(item, depth + 1) for key, item in value.items())
    return 1


def admit(raw: bytes) -> Any:
    if not raw or len(raw) > 1_048_576 or raw.startswith(b"\xef\xbb\xbf"):
        raise AdmissionError("byte limit or BOM")
    try:
        text = raw.decode("utf-8", "strict")
        value = json.loads(text, object_pairs_hook=_pairs, parse_int=_integer, parse_float=_number, parse_constant=lambda value: (_ for _ in ()).throw(AdmissionError(value)))
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise AdmissionError("invalid JSON or UTF-8") from error
    if _limits(value) > 200_000:
        raise AdmissionError("total value limit")
    return value


def _pointer(parts: Any) -> str:
    encoded = [str(part).replace("~", "~0").replace("/", "~1") for part in parts]
    return "/" + "/".join(encoded) if encoded else "/"


class Adapter:
    def __init__(self, repository_root: Path):
        resources: list[tuple[str, Resource[Any]]] = []
        self._schemas: dict[str, dict[str, Any]] = {}
        for path in sorted((repository_root / "contracts" / "schemas" / "v1").glob("*.schema.json")):
            raw = path.read_bytes()
            schema = json.loads(raw)
            metadata = schema["x-anvilkit-contract"]
            name = path.name.removesuffix(".schema.json")
            version = metadata["semanticVersion"]
            uri = f"anvilkit://schema/{name}.v{version.split('.', 1)[0]}@{version}?digest=sha256:{hashlib.sha256(raw).hexdigest()}"
            projected = dict(schema)
            projected["$schema"] = "https://json-schema.org/draft/2020-12/schema"
            projected["$id"] = uri
            projected.pop("x-anvilkit-contract", None)
            self._schemas[uri] = projected
            resources.append((uri, Resource.from_contents(projected)))
        self._registry = Registry().with_resources(resources)

    def validate(self, schema_uri: str, raw: bytes) -> list[Finding]:
        try:
            value = admit(raw)
        except AdmissionError:
            return [Finding("PARSE_REJECTED", "/", "/profile/strictAdmission")]
        schema = self._schemas.get(schema_uri)
        if schema is None:
            return [Finding("VALIDATION_FAILED", "/", "/profile/closedResolver")]
        validator = Draft202012Validator(schema, registry=self._registry, format_checker=FormatChecker())
        return sorted(Finding("VALIDATION_FAILED", _pointer(error.absolute_path), _pointer(error.absolute_schema_path)) for error in validator.iter_errors(value))

File: packages/contracts-go/validator/validator.go
// Package validator provides the pinned Go contract adapter. It performs strict
// byte admission before Draft 2020-12 validation and never resolves a network
// reference.
package validator

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"

	jsonschema "github.com/santhosh-tekuri/jsonschema/v6"
)

const maximumSafeInteger = 9007199254740991

const (
	maximumFindingCount = 100
	maximumSchemaBytes  = 1_048_576
	maximumSchemaCount  = 128
)

type Finding struct{ Code, InstancePath, SchemaPath string }

type Adapter struct{ schemas map[string]*jsonschema.Schema }

func logicalURI(name string, bytes []byte) string {
	digest := sha256.Sum256(bytes)
	return fmt.Sprintf("anvilkit://schema/%s?digest=sha256:%s", name, hex.EncodeToString(digest[:]))
}

func New(repositoryRoot string) (*Adapter, error) {
	directory := filepath.Join(repositoryRoot, "contracts", "agent", "schemas")
	schemaRoot, err := os.OpenRoot(directory)
	if err != nil {
		return nil, fmt.Errorf("open schema root: %w", err)
	}
	// The root is opened only to bound reads; closing it cannot fail in a
	// way that changes the schemas already loaded.
	defer func() { _ = schemaRoot.Close() }()
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, fmt.Errorf("read schema directory: %w", err)
	}
	if len(entries) > maximumSchemaCount {
		return nil, fmt.Errorf("schema directory exceeds %d entries", maximumSchemaCount)
	}
	compiler := jsonschema.NewCompiler()
	compiler.AssertFormat()
	compiler.UseLoader(denyLoader{})
	type source struct {
		uri   string
		value any
	}
	sources := make([]source, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".schema.json") {
			continue
		}
		if entry.Type()&os.ModeSymlink != 0 {
			return nil, fmt.Errorf("schema %s must not be a symbolic link", entry.Name())
		}
		info, err := entry.Info()
		if err != nil {
			return nil, fmt.Errorf("inspect %s: %w", entry.Name(), err)
		}
		if !info.Mode().IsRegular() || info.Size() > maximumSchemaBytes {
			return nil, fmt.Errorf("schema %s is not a regular bounded file", entry.Name())
		}
		raw, err := schemaRoot.ReadFile(entry.Name())
		if err != nil {
			return nil, fmt.Errorf("read %s: %w", entry.Name(), err)
		}
		admitted, err := Admit(raw)
		if err != nil {
			return nil, fmt.Errorf("parse %s: %w", entry.Name(), err)
		}
		value, ok := admitted.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("parse %s: schema root must be an object", entry.Name())
		}
		metadata, ok := value["x-anvilkit-contract"].(map[string]any)
		if !ok {
			return nil, fmt.Errorf("%s lacks metadata", entry.Name())
		}
		if _, ok := metadata["logicalId"].(string); !ok {
			return nil, fmt.Errorf("%s lacks a logical ID", entry.Name())
		}
		name := strings.TrimSuffix(entry.Name(), ".schema.json")
		uri := logicalURI(name, raw)
		value["$schema"] = "https://json-schema.org/draft/2020-12/schema"
		value["$id"] = uri
		delete(value, "x-anvilkit-contract")
		sources = append(sources, source{uri, value})
	}
	for _, source := range sources {
		if err := compiler.AddResource(source.uri, source.value); err != nil {
			return nil, fmt.Errorf("register %s: %w", source.uri, err)
		}
	}
	compiled := make(map[string]*jsonschema.Schema, len(sources))
	for _, source := range sources {
		schema, err := compiler.Compile(source.uri)
		if err != nil {
			return nil, fmt.Errorf("compile %s: %w", source.uri, err)
		}
		compiled[source.uri] = schema
	}
	return &Adapter{schemas: compiled}, nil
}

type denyLoader struct{}

func (denyLoader) Load(url string) (any, error) {
	return nil, fmt.Errorf("network/unknown schema resolution denied: %s", url)
}

func (a *Adapter) Validate(schemaURI string, raw []byte) []Finding {
	value, err := Admit(raw)
	if err != nil {
		return []Finding{{Code: "PARSE_REJECTED", InstancePath: "/", SchemaPath: "/profile/strictAdmission"}}
	}
	schema := a.schemas[schemaURI]
	if schema == nil {
		return []Finding{{Code: "VALIDATION_FAILED", InstancePath: "/", SchemaPath: "/profile/closedResolver"}}
	}
	if err := schema.Validate(value); err != nil {
		var validation *jsonschema.ValidationError
		if !errors.As(err, &validation) {
			return []Finding{{Code: "VALIDATION_FAILED", InstancePath: "/", SchemaPath: "/"}}
		}
		findings := make([]Finding, 0, maximumFindingCount)
		if flatten(validation, &findings, maximumFindingCount-1) {
			findings = append(findings, Finding{Code: "VALIDATION_FAILED", InstancePath: "/", SchemaPath: "/profile/findingLimit"})
		}
		sort.Slice(findings, func(i, j int) bool {
			if findings[i].Code != findings[j].Code {
				return findings[i].Code < findings[j].Code
			}
			if findings[i].InstancePath != findings[j].InstancePath {
				return findings[i].InstancePath < findings[j].InstancePath
			}
			return findings[i].SchemaPath < findings[j].SchemaPath
		})
		return findings
	}
	return nil
}

func pointer(parts []string) string {
	if len(parts) == 0 {
		return "/"
	}
	escaped := make([]string, len(parts))
	for i, part := range parts {
		escaped[i] = strings.ReplaceAll(strings.ReplaceAll(part, "~", "~0"), "/", "~1")
	}
	return "/" + strings.Join(escaped, "/")
}
func flatten(err *jsonschema.ValidationError, findings *[]Finding, limit int) bool {
	if len(*findings) >= limit {
		return true
	}
	if len(err.Causes) > 0 {
		for _, cause := range err.Causes {
			if flatten(cause, findings, limit) {
				return true
			}
		}
		return false
	}
	schemaPath := err.SchemaURL
	if keywordPath := err.ErrorKind.KeywordPath(); len(keywordPath) > 0 {
		schemaPath = strings.TrimSuffix(schemaPath, "#") + "#" + pointer(keywordPath)
	}
	*findings = append(*findings, Finding{Code: "VALIDATION_FAILED", InstancePath: pointer(err.InstanceLocation), SchemaPath: schemaPath})
	return false
}

func Admit(raw []byte) (any, error) {
	if len(raw) == 0 || len(raw) > 1_048_576 {
		return nil, errors.New("byte limit")
	}
	if !utf8.Valid(raw) || bytes.HasPrefix(raw, []byte{0xef, 0xbb, 0xbf}) {
		return nil, errors.New("invalid UTF-8/BOM")
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber()
	value, err := decodeValue(decoder, 0)
	if err != nil {
		return nil, err
	}
	if _, err = decoder.Token(); !errors.Is(err, io.EOF) {
		if err == nil {
			return nil, errors.New("trailing value")
		}
		return nil, err
	}
	return value, nil
}

func decodeValue(decoder *json.Decoder, depth int) (any, error) {
	if depth > 64 {
		return nil, errors.New("depth limit")
	}
	token, err := decoder.Token()
	if err != nil {
		return nil, err
	}
	switch value := token.(type) {
	case json.Delim:
		switch value {
		case '{':
			object := make(map[string]any)
			for decoder.More() {
				keyToken, err := decoder.Token()
				if err != nil {
					return nil, err
				}
				key, ok := keyToken.(string)
				if !ok {
					return nil, errors.New("object key")
				}
				if _, exists := object[key]; exists {
					return nil, errors.New("duplicate key")
				}
				child, err := decodeValue(decoder, depth+1)
				if err != nil {
					return nil, err
				}
				object[key] = child
			}
			end, err := decoder.Token()
			if err != nil || end != json.Delim('}') {
				return nil, errors.New("object end")
			}
			return object, nil
		case '[':
			array := make([]any, 0)
			for decoder.More() {
				if len(array) >= 100000 {
					return nil, errors.New("item limit")
				}
				child, err := decodeValue(decoder, depth+1)
				if err != nil {
					return nil, err
				}
				array = append(array, child)
			}
			end, err := decoder.Token()
			if err != nil || end != json.Delim(']') {
				return nil, errors.New("array end")
			}
			return array, nil
		}
	case json.Number:
		wire := value.String()
		if strings.HasPrefix(wire, "-") {
			number, err := strconv.ParseFloat(wire, 64)
			if err != nil || math.IsInf(number, 0) || number == 0 {
				return nil, errors.New("negative zero/range")
			}
		}
		if !strings.ContainsAny(wire, ".eE") {
			integer, err := strconv.ParseInt(wire, 10, 64)
			if err != nil || integer > maximumSafeInteger || integer < -maximumSafeInteger {
				return nil, errors.New("unsafe integer")
			}
			return integer, nil
		}
		number, err := strconv.ParseFloat(wire, 64)
		if err != nil || math.IsInf(number, 0) || number == 0 && wire != "0.0" {
			return nil, errors.New("number range")
		}
		return number, nil
	default:
		return token, nil
	}
	return nil, errors.New("invalid delimiter")
}
